src/transform/src/analysis/equivalences.rs

```diff
@@ -554,29 +559,55 @@ impl EquivalenceClasses {
 
     /// Produce the equivalences present in both inputs.
     pub fn union(&self, other: &Self) -> Self {
-        // TODO: seems like this could be extended, with similar concepts to localization:
-        //       We may removed non-shared constraints, but ones that remain could take over
-        //       and substitute in to retain more equivalences.
-
-        // For each pair of equivalence classes, their intersection.
-        let mut equivalences = EquivalenceClasses {
-            classes: Vec::new(),
-            remap: Default::default(),
-        };
-        for class1 in self.classes.iter() {
-            for class2 in other.classes.iter() {
-                let class = class1
-                    .iter()
-                    .filter(|e1| class2.iter().any(|e2| e1 == &e2))
-                    .cloned()
-                    .collect::<Vec<_>>();
-                if class.len() > 1 {
-                    equivalences.classes.push(class);
+        self.union_many(&[other])
+    }
+
+    /// The equivalence classes of terms equivalent in all inputs.
+    ///
+    /// This method currently misses opportunities, because it only looks for exactly matches in expressions,
+    /// which may not include all possible matches. For example, `f(#1) == g(#1)` may exist in one class, but
+    /// in another class where `#0 == #1` it may exist as `f(#0) == g(#0)`.
+    pub fn union_many(&self, others: &[&Self]) -> Self {
+        if others.is_empty() {
+            self.clone()
+        } else {
+            // Map from expression to an equivalence class identifier.
+            // That identifier is a pair of integer and expression reference, where
+            // the integer is a bit of a "hashcons" of accumulated class representatives.
+            let mut intersection: BTreeMap<&MirScalarExpr, usize> = Default::default();
+            let mut rekey: BTreeMap<_, _> = Default::default();
+            for (key, val) in self.remap.iter() {
```

Comment:
It depends what you mean by "safe", but .. it has the same type of defect that the prior code has, which is that if the two inputs are in a "stale" state, e.g. `classes` is not an equivalence relation, then the output is not the intersection of the equivalence classes represented by the inputs. This does a specific thing, intersect the `remap` members of each input, whereas the prior implementation does an unspecified thing. This is the thing called out by the last paragraph of the PR description.

Reply:
There is [this comment](https://github.com/MaterializeInc/materialize/pull/30134#pullrequestreview-2395440828) with an idea on how to solve this defect. What do you think of that idea? My worry is that even if the current code always happens to call it at a correct moment, this is a bit of a footgun during future code changes (if I understand the problem correctly).

Reply:
Right, so the comment doesn't work as stated, because you get `&Self` arguments as input and cannot `refresh()` from a borrowed reference. We could .. clone them in that case and refresh the clone and use it, we could put everything behind a `RefCell`, maybe some other things. Mostly, it feels a bit disproportionate as the concern is "does the new, very specific behavior deviate too much from the prior non-specific behavior", which the dirty bit would not address.

Reply:
The concern is valid, but I think the same amount of valid as "what if I ask for a `reducer()` without having called `refresh()` or `minimize()` first?". You get the answer that corresponds to the last time you refreshed, without the opinion that you must refresh before you are allowed to use the type. This is similar to what `egg` does, which is allow an accumulation of work without preventing read access, though .. I don't think we have the same justification (a high volume of e-matching).

It's easy enough to make fields private, and install a `todo` queue, and the "dirty bit" is "is the queue empty", but I don't think we know enough yet to proscribe using the type while there is anything in the todo list.

Reply:
I started this but need to pick it up later. The problem is that at least one place in EqProp we need to mutably iterate over the members of `Equivalence::classes`, and short of writing a method for "do random things to `classes` but I'll call `minimize()` after for this one case" it's fairly grotty.

Reply:
On the train I concluded that the relevant bit of code can be improved, which .. is always a risk. Specifically, the relevant code in EqProp is testing the outer and inner equivalences to see if they are contradictory, which would indicate that the relation expression can be pruned. It does this by reducing the terms of outer via inner, and minimizing, but this has the potential to miss things. For example
```
inner: [[Lit(0), Col(0)], [Col(1), Col(2)]]
outer: [[Lit(1), Col(2)], [Col(0), Col(1)]]
```
you actually need to put all these classes together and minimize to see the contradiction `Lit(0) = Lit(1)`, which we don't do. When you do that, you don't need mutable access to the members of `outer` because you are going to clone the crud out of everything anyhow.

Related observations:
1. The `todo` member has the potential to be a perf improvement, because when you have to ask to insert equivalences, we can quickly check that they novel, and if not we don't enqueue them, and don't "dirty" the data structure. No clue how often this happens, but it could.
2. We `reduce` expressions a bunch in EqProp and they should probably be `reduce_safely`, to avoid propagating error woes further. I'll add that back in somewhere, but probably not this PR.

Imo, the above are all good ideas, but increasingly non-performance behavior changing. We can see what happens, but if there is any non-trivial outcome changes, I don't think folding them in to this PR is the right thing. We can always do a software engineering follow-up where we provide a more error-proof interface to `Equivalences`, and see if it has planning fallout, as a separate piece of work.

Reply:
Hmm, ok, this is indeed getting a bit complex. I'm totally ok with merging the PR, and then thinking about follow-ups.

Reply:
Oh, now there seem to be some slt fails.

Reply:
They are covered by the "we can see what happens" clause: it seems that tidying up the code and the resulting behavioral change is non-trivial.

Reply:
I'm on a different PR at the moment, but can pop off the last commit to return to green, and put it in a follow-up PR that we can study to see what is going on (short look indicates that not all changes are especially clean).

Reply:
Other commit is up in draft at https://github.com/MaterializeInc/materialize/pull/30272